File: native/src/openal_bridge.c
#include "openal_bridge.h"
#include "compat_runtime.h"
#include <OpenAL/al.h>
#include <OpenAL/alc.h>
#include <pthread.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#pragma clang diagnostic ignored "-Wdeprecated-declarations"

/* Host contexts/devices never fit into guest pointer slots. AL buffer/source
 * names and sample data have the same widths in both ABIs. */
static struct { void *pointer; uint32_t token; bool context; } objects[1024];
static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
static uint32_t next_token = 0x7f140000;
static uint32_t token(void *p, bool context) {
    if (!p) return 0;
    pthread_mutex_lock(&lock);
    unsigned empty = 1024;
    for (unsigned i=0;i<1024;++i) {
        if (objects[i].pointer==p && objects[i].context==context) {
            uint32_t t=objects[i].token; pthread_mutex_unlock(&lock); return t;
        }
        if (!objects[i].pointer && empty==1024) empty=i;
    }
    uint32_t t=0;
    if(empty<1024) { t=next_token++; objects[empty].pointer=p; objects[empty].token=t; objects[empty].context=context; }
    pthread_mutex_unlock(&lock); return t;
}
static void *pointer(uint32_t t, bool context, bool remove) {
    void *p=NULL; pthread_mutex_lock(&lock);
    for(unsigned i=0;i<1024;++i) if(objects[i].token==t && objects[i].context==context) {
        p=objects[i].pointer; if(remove) memset(&objects[i],0,sizeof(objects[i])); break;
    }
    pthread_mutex_unlock(&lock); return p;
}
static bool muted(void) { const char *v=getenv("LP32_MUTE_AUDIO"); return v && strcmp(v,"0"); }
int openal_bridge32_dispatch(const char *name,const uint32_t *a,uint64_t *out) {
#define IS(n) (!strcmp(name,n))
#define P(i) ((void *)(uintptr_t)a[i])
    *out=0;
    if(IS("_alcOpenDevice")) *out=token(alcOpenDevice(P(0)),false);
    else if(IS("_alcCloseDevice")) {
        ALCdevice *d=pointer(a[0],false,false); *out=alcCloseDevice(d);
        if(*out) pointer(a[0],false,true);
    } else if(IS("_alcCreateContext")) *out=token(alcCreateContext(pointer(a[0],false,false),P(1)),true);
    else if(IS("_alcDestroyContext")) alcDestroyContext(pointer(a[0],true,true));
    else if(IS("_alcGetCurrentContext")) *out=token(alcGetCurrentContext(),true);
    else if(IS("_alcMakeContextCurrent")) {
        *out=alcMakeContextCurrent(pointer(a[0],true,false));
        if(*out && a[0] && muted()) alListenerf(AL_GAIN,0);
    } else if(IS("_alcGetEnumValue")) *out=alcGetEnumValue(pointer(a[0],false,false),P(1));
    else if(IS("_alcIsExtensionPresent")) *out=alcIsExtensionPresent(pointer(a[0],false,false),P(1));
    else if(IS("_alIsExtensionPresent")) *out=alIsExtensionPresent(P(0));
    else if(IS("_alcGetProcAddress") || IS("_alGetProcAddress")) {
        const char *s=P(IS("_alcGetProcAddress")?1:0);
        if(s && !strcmp(s,"alBufferDataStatic")) *out=compat_runtime32_guest_callback("_alBufferDataStatic");
    } else if(IS("_alGenBuffers")) alGenBuffers((int)a[0],P(1));
    else if(IS("_alGenSources")) alGenSources((int)a[0],P(1));
    else if(IS("_alDeleteBuffers")) alDeleteBuffers((int)a[0],P(1));
    else if(IS("_alDeleteSources")) alDeleteSources((int)a[0],P(1));
    else if(IS("_alBufferData") || IS("_alBufferDataStatic")) alBufferData(a[0],a[1],P(2),a[3],a[4]);
    else if(IS("_alDistanceModel")) alDistanceModel(a[0]);
    else if(IS("_alGetError")) *out=alGetError();
    else if(IS("_alGetFloat")) *out=compat_runtime32_return_double(alGetFloat(a[0]));
    else if(IS("_alGetSourcei")) alGetSourcei(a[0],a[1],P(2));
    else if(IS("_alListenerfv")) {
        float zero=0; alListenerfv(a[0],muted() && a[0]==AL_GAIN ? &zero : P(1));
    } else if(IS("_alSourcePlay")) { if(muted())alListenerf(AL_GAIN,0); alSourcePlay(a[0]); }
    else if(IS("_alSourceStop")) alSourceStop(a[0]);
    else if(IS("_alSourceQueueBuffers")) alSourceQueueBuffers(a[0],a[1],P(2));
    else if(IS("_alSourceUnqueueBuffers")) alSourceUnqueueBuffers(a[0],a[1],P(2));
    else if(IS("_alSourcei")) alSourcei(a[0],a[1],a[2]);
    else if(IS("_alSourcefv")) alSourcefv(a[0],a[1],P(2));
    else if(IS("_alSourcef")) { float f;memcpy(&f,a+2,4);alSourcef(a[0],a[1],f); }
    else return 0;
    return 1;
}

File: native/src/openal_bridge.m
#import <Foundation/Foundation.h>
#import <OpenAL/al.h>
#import <OpenAL/alc.h>
#include "openal_bridge.h"
#include "objc_bridge.h"
#include "compat_runtime.h"
#include <string.h>

#pragma clang diagnostic ignored "-Wdeprecated-declarations"

static void *pointer(uint32_t value) { return [(NSValue *)objc_bridge32_host_object(value) pointerValue]; }
static uint32_t handle(void *value) { return value ? objc_bridge32_guest_object([NSValue valueWithPointer:value]) : 0; }
static float real(uint32_t bits) { float value; memcpy(&value, &bits, 4); return value; }
static bool muted(void) { return getenv("LP32_MUTE_AUDIO") || getenv("LP32_HEADLESS"); }
static uint32_t copy_string(const char *text, bool list) {
    if (!text) return 0;
    size_t size = strlen(text) + 1;
    if (list) { while (text[size]) size += strlen(text + size) + 1; ++size; }
    static NSMutableDictionary *strings;
    static NSLock *lock;
    static dispatch_once_t once;
    dispatch_once(&once, ^{ strings = [[NSMutableDictionary alloc] init]; lock = [[NSLock alloc] init]; });
    NSData *key = [NSData dataWithBytes:text length:size];
    [lock lock]; uint32_t result = [strings[key] unsignedIntValue];
    if (!result && size <= UINT32_MAX) {
        result = compat_runtime32_allocate((uint32_t)size, 0);
        if (result) { memcpy((void *)(uintptr_t)result, text, size); strings[key] = @(result); }
    }
    [lock unlock]; return result;
}

int openal_bridge32_dispatch(const char *name, const uint32_t *a, uint64_t *result) {
    if (strncmp(name, "_al", 3)) return 0;
    @autoreleasepool {
#define IS(s) (!strcmp(name, "_" s))
#define PTR(i) ((void *)(uintptr_t)a[i])
#define RETURN(v) do { *result = (uint64_t)(v); return 1; } while (0)
    if (IS("alcOpenDevice")) RETURN(handle(alcOpenDevice(PTR(0))));
    if (IS("alcCloseDevice")) RETURN(alcCloseDevice(pointer(a[0])));
    if (IS("alcCreateContext")) RETURN(handle(alcCreateContext(pointer(a[0]), PTR(1))));
    if (IS("alcDestroyContext")) { alcDestroyContext(pointer(a[0])); RETURN(0); }
    if (IS("alcMakeContextCurrent")) {
        ALCboolean ok = alcMakeContextCurrent(pointer(a[0]));
        if (ok && a[0] && muted()) alListenerf(AL_GAIN, 0);
        RETURN(ok);
    }
    if (IS("alcProcessContext")) { alcProcessContext(pointer(a[0])); RETURN(0); }
    if (IS("alcSuspendContext")) { alcSuspendContext(pointer(a[0])); RETURN(0); }
    if (IS("alcGetError")) RETURN(alcGetError(pointer(a[0])));
    if (IS("alcGetString")) RETURN(copy_string(alcGetString(pointer(a[0]), a[1]), !a[0] && a[1] == ALC_DEVICE_SPECIFIER && alcIsExtensionPresent(NULL, "ALC_ENUMERATION_EXT")));
    if (IS("alGetError")) RETURN(alGetError());
    if (IS("alGetString")) RETURN(copy_string(alGetString(a[0]), false));
    if (IS("alGenBuffers")) { alGenBuffers(a[0], PTR(1)); RETURN(0); }
    if (IS("alDeleteBuffers")) { alDeleteBuffers(a[0], PTR(1)); RETURN(0); }
    if (IS("alGenSources")) { alGenSources(a[0], PTR(1)); RETURN(0); }
    if (IS("alDeleteSources")) { alDeleteSources(a[0], PTR(1)); RETURN(0); }
    if (IS("alBufferData")) { alBufferData(a[0], a[1], PTR(2), a[3], a[4]); RETURN(0); }
    if (IS("alSourceQueueBuffers")) { alSourceQueueBuffers(a[0], a[1], PTR(2)); RETURN(0); }
    if (IS("alSourceUnqueueBuffers")) { alSourceUnqueueBuffers(a[0], a[1], PTR(2)); RETURN(0); }
    if (IS("alSourcePlay")) { alSourcePlay(a[0]); RETURN(0); }
    if (IS("alSourcePause")) { alSourcePause(a[0]); RETURN(0); }
    if (IS("alGetSourcef")) { alGetSourcef(a[0], a[1], PTR(2)); RETURN(0); }
    if (IS("alGetSourcei")) { alGetSourcei(a[0], a[1], PTR(2)); RETURN(0); }
    if (IS("alSourcef")) { alSourcef(a[0], a[1], real(a[2])); RETURN(0); }
    if (IS("alSourcei")) { alSourcei(a[0], a[1], a[2]); RETURN(0); }
    if (IS("alSource3f")) { alSource3f(a[0], a[1], real(a[2]), real(a[3]), real(a[4])); RETURN(0); }
    if (IS("alListener3f")) { alListener3f(a[0], real(a[1]), real(a[2]), real(a[3])); RETURN(0); }
    if (IS("alListenerf")) { alListenerf(a[0], a[0] == AL_GAIN && muted() ? 0 : real(a[1])); RETURN(0); }
    if (IS("alListenerfv")) {
        if (a[0] == AL_GAIN && muted()) alListenerf(AL_GAIN, 0);
        else alListenerfv(a[0], PTR(1));
        RETURN(0);
    }
    if (IS("alDistanceModel")) { alDistanceModel(a[0]); RETURN(0); }
    if (IS("alDopplerFactor")) { alDopplerFactor(real(a[0])); RETURN(0); }
    if (IS("alDopplerVelocity")) { alDopplerVelocity(real(a[0])); RETURN(0); }
    if (IS("alSpeedOfSound")) { alSpeedOfSound(real(a[0])); RETURN(0); }
    }
    return 0;
#undef IS
#undef PTR
#undef RETURN
}

int openal_bridge32_self_test(void) {
    @autoreleasepool {
        uint32_t memory = compat_runtime32_allocate(4096, 1);
        if (!memory) return -1;
        uint32_t a[8] = {0}; uint64_t r = 0;
#define CALL(s) openal_bridge32_dispatch("_" s, a, &r)
        CALL("alcOpenDevice"); uint32_t device = r;
        if (!device) { compat_runtime32_deallocate(memory); return -1; }
        a[0] = device; a[1] = 0; CALL("alcCreateContext"); uint32_t context = r;
        int failed = !context;
        if (context) {
            a[0] = context; CALL("alcMakeContextCurrent"); failed |= !r;
            float listener_gain = .75f, actual_gain = -1;
            a[0] = AL_GAIN; memcpy(a + 1, &listener_gain, 4); CALL("alListenerf");
            alGetListenerf(AL_GAIN, &actual_gain); failed |= actual_gain != (muted() ? 0 : listener_gain);
            memcpy((void *)(uintptr_t)(memory + 16), &listener_gain, 4);
            a[1] = memory + 16; CALL("alListenerfv");
            alGetListenerf(AL_GAIN, &actual_gain); failed |= actual_gain != (muted() ? 0 : listener_gain);
            a[0] = 1; a[1] = memory; CALL("alGenBuffers"); uint32_t buffer = *(uint32_t *)(uintptr_t)memory;
            a[1] = memory + 4; CALL("alGenSources"); uint32_t source = *(uint32_t *)(uintptr_t)(memory + 4);
            a[0] = buffer; a[1] = AL_FORMAT_MONO16; a[2] = memory + 32; a[3] = 1024; a[4] = 22050; CALL("alBufferData");
            a[0] = source; a[1] = AL_BUFFER; a[2] = buffer; CALL("alSourcei");
            a[2] = memory + 8; CALL("alGetSourcei"); failed |= *(uint32_t *)(uintptr_t)(memory + 8) != buffer;
            float gain = .375f; a[1] = AL_GAIN; memcpy(a + 2, &gain, 4); CALL("alSourcef");
            a[2] = memory + 12; CALL("alGetSourcef"); failed |= *(float *)(uintptr_t)(memory + 12) != gain;
            CALL("alGetError"); failed |= r != AL_NO_ERROR;
            a[0] = 1; a[1] = memory + 4; CALL("alDeleteSources");
            a[1] = memory; CALL("alDeleteBuffers");
            a[0] = 0; CALL("alcMakeContextCurrent"); a[0] = context; CALL("alcDestroyContext");
        }
        a[0] = device; CALL("alcCloseDevice"); failed |= !r;
        compat_runtime32_deallocate(memory);
        fprintf(stderr, "OpenAL self-test: %s buffer upload, source attachment, float parameters, cleanup (no playback)\n", failed ? "FAIL" : "PASS");
        return failed ? -1 : 0;
#undef CALL
    }
}
